psycopg/psycopg/types/composite.py

```diff
@@ -219,63 +223,136 @@ def __init__(self, oid: int, context: abc.AdaptContext | None = None):
         # records), oids might differ and we'd need separate transformers.
         self._txs: dict[tuple[int, ...], abc.Transformer] = {}
 
-    def load(self, data: abc.Buffer) -> tuple[Any, ...]:
+    def _parse_record(self, data: abc.Buffer) -> tuple[Any, ...]:
         nfields = unpack_len(data, 0)[0]
+        if not nfields:
+            return ()
+
+        oids, record = zip(*self._iterunpack(data, nfields))
+        tx = self._get_transformer(oids)
+        return tx.load_sequence(record)
+
+    @staticmethod
+    def _iterunpack(
+        data: abc.Buffer, nfields: int
+    ) -> Iterator[tuple[int, abc.Buffer | None]]:
```

Comment:
If you did this as an optimization, I doubt it is. You are creating several intermediate tuples and call into the iterator protocol, which requires throwing and catching an exception to terminate.

If you can make a micro-benchmark to prove that this is better than building two lists I can accept this changeset, otherwise it would be better to leave it as it was.

Actually, a micro-optimization would be to preallocate the lists, as we know the lengths, and set the items. Could you maybe try if this doesn't cause a regression and which solution is faster?

```python
oids = [0] * nfields
record = [None] * nfields
for i in range(nfields):
    oids[i], length = _unpack_oidlen(data, offset)
    offset += 8
    if length >= 0:
        record[i] = data[offset : offset + length]
        offset += length
```

Reply:
In general I try to avoid `list.append` because it forces the interpreter to check if more memory needs be allocated with every call (not to mention the cost if it does, in fact, need to allocate more memory).

I can compare the two suggestions - I agree creating intermediate tuples incurs its own cost, but tuples are fairly well-optimize and pre-allocating the list is likely to be the winner here.